File: frontend/src/types/activity.ts
import type { JlptLevel } from './profile'

export type ActivityCategory = 'vocabulary' | 'grammar' | 'kanji'

export interface VocabularyItem {
  id: string
  term: string
  reading: string
  meaning: string
  level: JlptLevel
  example_sentence: string | null
  example_translation: string | null
}

export interface GrammarConcept {
  id: string
  key: string
  title: string
  level: JlptLevel
  explanation: string
  example_sentence: string
  example_translation: string
  answer: string
}

export interface KanjiItem {
  id: string
  character: string
  onyomi: string
  kunyomi: string
  meaning: string
  level: JlptLevel
  example_word: string | null
  example_reading: string | null
  example_meaning: string | null
}

export interface QuizQuestion {
  item_id: string
  prompt: string
  options: string[]
}

export interface QuizResponse {
  category: ActivityCategory
  level: JlptLevel
  questions: QuizQuestion[]
}

export interface QuizAnswer {
  item_id: string
  selected: string
}

export interface QuizSubmitPayload {
  category: ActivityCategory
  level: JlptLevel
  answers: QuizAnswer[]
}

export interface QuizResultItem {
  item_id: string
  correct: boolean
  correct_answer: string
}

export interface QuizSubmitResponse {
  correct_count: number
  total: number
  xp_earned: number
  total_xp: number
  results: QuizResultItem[]
}

export interface FlashcardReview {
  item_id: string
  known: boolean
}

export interface FlashcardCompletePayload {
  category: ActivityCategory
  level: JlptLevel
  reviewed: FlashcardReview[]
}

export interface FlashcardCompleteResponse {
  known_count: number
  total: number
  xp_earned: number
  total_xp: number
}

export interface ActivityHistoryEntry {
  category: ActivityCategory
  activity_type: string
  level: JlptLevel
  correct_count: number | null
  known_count: number | null
  total: number
  xp_earned: number
  created_at: string
}

File: backend/app/schemas/activity.py
from datetime import datetime
from enum import Enum
from typing import Any

from pydantic import BaseModel, Field

from app.schemas.profile import JLPTLevel


class ActivityCategory(str, Enum):
    VOCABULARY = "vocabulary"
    GRAMMAR = "grammar"
    KANJI = "kanji"


class QuizQuestion(BaseModel):
    item_id: str
    prompt: str
    options: list[str]


class QuizResponse(BaseModel):
    category: ActivityCategory
    level: JLPTLevel
    questions: list[QuizQuestion]


class QuizAnswer(BaseModel):
    item_id: str
    selected: str


class QuizSubmitRequest(BaseModel):
    category: ActivityCategory
    level: JLPTLevel
    answers: list[QuizAnswer] = Field(min_length=1)


class QuizResultItem(BaseModel):
    item_id: str
    correct: bool
    correct_answer: str


class QuizSubmitResponse(BaseModel):
    correct_count: int
    total: int
    xp_earned: int
    total_xp: int
    results: list[QuizResultItem]


class FlashcardReview(BaseModel):
    item_id: str
    known: bool


class FlashcardCompleteRequest(BaseModel):
    category: ActivityCategory
    level: JLPTLevel
    reviewed: list[FlashcardReview] = Field(min_length=1)


class FlashcardCompleteResponse(BaseModel):
    known_count: int
    total: int
    xp_earned: int
    total_xp: int


class ActivityHistoryEntry(BaseModel):
    category: ActivityCategory
    activity_type: str
    level: JLPTLevel
    correct_count: int | None = None
    known_count: int | None = None
    total: int
    xp_earned: int
    created_at: datetime


def activity_to_history_entry(doc: dict[str, Any]) -> ActivityHistoryEntry:
    return ActivityHistoryEntry(
        category=doc["category"],
        activity_type=doc["activity_type"],
        level=doc["level"],
        correct_count=doc.get("correct_count"),
        known_count=doc.get("known_count"),
        total=doc["total"],
        xp_earned=doc["xp_earned"],
        created_at=doc["created_at"],
    )
